pageserver/src/tenant/storage_layer/delta_layer.rs

```diff
@@ -994,94 +988,76 @@ impl DeltaLayerInner {
         reconstruct_state: &mut ValuesReconstructState,
         ctx: &RequestContext,
     ) {
-        let vectored_blob_reader = VectoredBlobReader::new(&self.file);
-        let mut ignore_key_with_err = None;
-
         let max_vectored_read_bytes = self
             .max_vectored_read_bytes
             .expect("Layer is loaded with max vectored bytes config")
             .0
             .into();
         let buf_size = Self::get_min_read_buffer_size(&reads, max_vectored_read_bytes);
-        let mut buf = Some(BytesMut::with_capacity(buf_size));
+
+        let mut ignore_key_with_err = None;
 
         // Note that reads are processed in reverse order (from highest key+lsn).
         // This is the order that `ReconstructState` requires such that it can
         // track when a key is done.
         for read in reads.into_iter().rev() {
-            let res = vectored_blob_reader
-                .read_blobs(&read, buf.take().expect("Should have a buffer"), ctx)
-                .await;
-
-            let blobs_buf = match res {
-                Ok(blobs_buf) => blobs_buf,
-                Err(err) => {
-                    let kind = err.kind();
-                    for (_, blob_meta) in read.blobs_at.as_slice() {
-                        reconstruct_state.on_key_error(
-                            blob_meta.key,
-                            PageReconstructError::Other(anyhow!(
-                                "Failed to read blobs from virtual file {}: {}",
-                                self.file.path(),
-                                kind
-                            )),
-                        );
-                    }
-
-                    // We have "lost" the buffer since the lower level IO api
-                    // doesn't return the buffer on error. Allocate a new one.
-                    buf = Some(BytesMut::with_capacity(buf_size));
+            let mut senders: HashMap<
+                (Key, Lsn),
+                sync::oneshot::Sender<Result<OnDiskValue, std::io::Error>>,
+            > = Default::default();
+            for (_, blob_meta) in read.blobs_at.as_slice().iter().rev() {
+                let (tx, rx) = sync::oneshot::channel();
+                senders.insert((blob_meta.key, blob_meta.lsn), tx);
+                reconstruct_state.update_key(
+                    &blob_meta.key,
+                    blob_meta.lsn,
+                    blob_meta.will_init,
+                    rx,
+                );
```

Comment:
> Is there a compelling reason why we chose to use channels here instead of making update_key an async function and passing it an IO future for the value?

`update_key` shouldn't wait on the completion of the IO, so I don't really understand why we would make it async.
The receiver is just pushed into `VectoredValueReconstructState::on_disk_values` and that vector is drained at the end of the read in `VectoredValueReconstructState::collect_pending_ios`.

I guess we could change `VectoredValueReconstructState::on_disk_values` to stash IO futures instead of oneshot channels, but I don't really see what that gets us.

Reply:
> update_key shouldn't wait on the completion of the IO, so I don't really understand why we would make it async.

I think I just intuitively considered it easier to understand updating the state once the read completed, which we could do by nesting futures, but I don't think it has any practical consequence. We'd probably run into some ownership issues as well.

>I guess we could change VectoredValueReconstructState::on_disk_values to stash IO futures instead of oneshot channels, but I don't really see what that gets us.

I think there are a few benefits to consider:

* Spawning a task has a cost, and spawning many tasks puts pressure on the Tokio scheduler. I don't know if this matters considering we also do IO here, but it might at scale. Futures can be awaited on the current task without spawning new ones.

* Serial/parallel operation is trivial simply by awaiting the futures sequentially (serial) or using e.g. `FuturesOrdered` or `FuturesUnordered` to await them all concurrently on the same task (parallel) -- the latter also optimizes away polling costs.

* Futures are automatically cancelled when dropped, while spawned tasks need explicit cancellation (you alluded to this being a problem in the PR description).

* We avoid the additional channel synchronization and overhead. This probably doesn't matter considering we're doing IO.

Reply:
> I think I just intuitively considered it easier to understand updating the state once the read completed, which we could do by nesting futures, but I don't think it has any practical consequence. We'd probably run into some ownership issues as well.

Yeah, I agree. The issue with that is that we don't get concurrency. The gist here is that we move on to the next layer after issuing IO without waiting on completion.

> I think there are a few benefits to consider

That's fair. I'll give it a go when I get the time.

Reply:
> Yeah, I agree. The issue with that is that we don't get concurrency. The gist here is that we move on to the next layer after issuing IO without waiting on completion.

I meant that these would just wrap the IO future, but not await it straight away -- we'd await the futures concurrently. The state would update as each read completed. But as I said, this might cause ownership issues (we might get away with it if we're not spawning tasks, but could get ugly either way).

Reply:
I gave this a go. It's doable but quite tricky. The difficulty comes from the fact that we need to fan out futures from the IO future itself (which reads data for multiple keys). This would be quite inefficient currently because we have to iterate over all `VectoredBlobsBuff::blobs` for each blob.

I think we should defer this to when we pick this back up and benchmark it. The task and future approaches should be benchmarked separately. With the task approach we actually perform IO while the traversal is happening, but with the futures approach all IO would be done at the end.

Reply:
Yeah, benchmarks seem reasonable. I'm a bit worried that putting this amount of pressure on the Tokio scheduler could cause latency issues, probably more than the gains we get from executing IOs earlier, but we can find out with appropriate benchmarks.